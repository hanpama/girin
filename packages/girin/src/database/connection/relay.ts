import { MetadataStorage, ObjectType, Field, TypeExpression, List, NonNull, StringScalar } from "girin-typelinker";
import { Connection, Edge } from "./base";


export function loadPredefinedConnectionFields(storage: MetadataStorage) {
  const pageInfoField = new Field({
    args: [],
    defaultName: 'pageInfo',
    type: new TypeExpression('PageInfo'),
  });
  const cursorField = new Field({
    args: [],
    defaultName: 'cursor',
    type: NonNull.of(StringScalar),
  });

  storage.registerFieldReference({
    field: pageInfoField,
    definitionClass: Connection,
  });
  storage.registerFieldReference({
    field: cursorField,
    definitionClass: Edge,
  });
}

export function defineConnection(target: Function, storage: MetadataStorage) {
  const definition = storage.getDefinition(ObjectType, target, 'any');
  const typeName = definition.metadata.typeName;

  const connectionTypeName = `${typeName}Connection`;
  const edgeTypeName = `${typeName}Edge`;

  const connectionType = new ObjectType({
    typeName: connectionTypeName,
    description: '',
  });
  const edgeType = new ObjectType({
    typeName: edgeTypeName,
    description: '',
  });

  const edgesField = new Field({
    args: [],
    defaultName: 'edges',
    type: List.of(target),
  });
  const nodeField = new Field({
    args: [],
    defaultName: 'node',
    type: new TypeExpression(target)
  });

  storage.register(connectionType, Connection);
  storage.register(edgeType, Edge);
  storage.registerExtensionFieldReference({ field: edgesField, extendingTypeName: connectionTypeName });
  storage.registerExtensionFieldReference({ field: nodeField, extendingTypeName: edgeTypeName });

  return {
    connectionType: new TypeExpression(connectionTypeName),
    edgeType: new TypeExpression(edgeTypeName),
  };
}