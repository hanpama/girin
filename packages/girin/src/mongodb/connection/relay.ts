// import {
//   MetadataStorage,
//   ObjectType,
//   Field,
//   TypeExpression,
//   List,
//   NonNull,
//   StringScalar,
//   FieldReferenceEntry,
//   DefinitionEntry,
//   FieldMixinEntry,
// } from "@girin/typelink";

// import { Connection, Edge } from "./base";


// export function loadPredefinedConnectionFields(storage: MetadataStorage) {
//   const pageInfoField = new Field({
//     args: [],
//     defaultName: 'pageInfo',
//     type: new TypeExpression('PageInfo'),
//   });
//   const cursorField = new Field({
//     args: [],
//     defaultName: 'cursor',
//     type: NonNull.of(StringScalar),
//   });

//   storage.registerEntry(Connection, new FieldReferenceEntry({ field: pageInfoField }));
//   storage.registerEntry(Edge, new FieldReferenceEntry({ field: cursorField }));
// }

// export function defineConnection(target: Function, storage: MetadataStorage) {
//   const definition = storage.getDefinition(ObjectType, target, 'any');
//   const typeName = definition.metadata.typeName;

//   const connectionTypeName = `${typeName}Connection`;
//   const edgeTypeName = `${typeName}Edge`;

//   const connectionType = new ObjectType({
//     typeName: connectionTypeName,
//     description: '',
//   });
//   const edgeType = new ObjectType({
//     typeName: edgeTypeName,
//     description: '',
//   });

//   const edgesField = new Field({
//     args: [],
//     defaultName: 'edges',
//     type: List.of(target),
//   });
//   const nodeField = new Field({
//     args: [],
//     defaultName: 'node',
//     type: new TypeExpression(target)
//   });

//   storage.registerEntry(Connection, new DefinitionEntry({ metadata: connectionType }));
//   storage.registerEntry(Connection, new FieldMixinEntry({ field: edgesField, extendingTypeName: connectionTypeName }));
//   storage.registerEntry(Edge, new DefinitionEntry({ metadata: edgeType }));
//   storage.registerEntry(Edge, new FieldMixinEntry({ field: nodeField, extendingTypeName: edgeTypeName }));

//   return {
//     connectionType: new TypeExpression(connectionTypeName),
//     edgeType: new TypeExpression(edgeTypeName),
//   };
// }