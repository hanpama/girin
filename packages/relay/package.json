{
  "name": "@girin/relay",
  "version": "0.5.0-alpha.4",
  "description": "Relay types for Girin",
  "main": "dist",
  "types": "dist/index.d.ts",
  "scripts": {
    "prepublish": "npm run build",
    "build": "npm run compile",
    "compile": "tsc"
  },
  "keywords": [
    "graphql"
  ],
  "license": "MIT",
  "author": "Max Choi <hanpama@gmail.com>",
  "tags": [
    "graphql"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/hanpama/girin.git"
  },
  "bugs": {
    "url": "https://github.com/hanpama/girin/issues"
  },
  "homepage": "https://github.com/hanpama/girin",
  "dependencies": {
    "@girin/typelink": "^0.5.0-alpha.4"
  },
  "gitHead": "7f87fc9641454bbe0e6367e5b1f15eea4a6fa956"
}
