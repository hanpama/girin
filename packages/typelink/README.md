# Girin: typelink



### Starwars Schema

* [Starwars Schema](https://github.com/hanpama/girin/tree/master/packages/girin/test/starwars/starWarsSchema.ts)
* [Starwars Schema Relay](https://github.com/hanpama/girin/tree/master/packages/girin-relay/test/starWarsSchema.ts)
