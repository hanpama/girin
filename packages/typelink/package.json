{
  "name": "@girin/typelink",
  "version": "0.5.0-alpha.4",
  "description": "Define GraphQL types linked to classes",
  "author": "Max Choi <hanpama@gmail.com>",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "tags": [
    "graphql"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/hanpama/girin.git"
  },
  "bugs": {
    "url": "https://github.com/hanpama/girin/issues"
  },
  "homepage": "https://github.com/hanpama/girin",
  "license": "MIT",
  "peerDependencies": {
    "graphql": "^0.11.0 || ^0.12.0 || ^0.13.0 || ^14.0.0"
  },
  "gitHead": "7f87fc9641454bbe0e6367e5b1f15eea4a6fa956",
  "files": [
    "./lib"
  ],
  "scripts": {
    "build": "tsc -b .",
    "prepublish": "npm run build"
  }
}