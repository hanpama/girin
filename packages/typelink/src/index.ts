export * from './sdl';
export * from './definition';
export * from './field';
export * from './global';
export * from './type-expression';
export * from './metadata';
