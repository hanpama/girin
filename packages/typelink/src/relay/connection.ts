import { GraphQLObjectType, GraphQLBoolean, GraphQLString, GraphQLNonNull, defaultFieldResolver } from 'graphql';

import {
  TypeArg, TypeExpression, type, List,
  getGlobalMetadataStorage,
  Field, ObjectType
} from '..';
import { bindStaticResolver } from '../sdl/ast';


type ConnectionByNode = { node: TypeArg | TypeExpression, edge?: undefined };
type ConnectionByEdge = { edge: TypeArg | TypeExpression, node?: undefined };
type DefineConnectionOptions = { typeName?: string } & (ConnectionByNode | ConnectionByEdge);


export function defineConnection(options: DefineConnectionOptions) {
  return function registerConnectionMetadata(definitionClass: Function) {
    const storage = getGlobalMetadataStorage();

    storage.deferRegister(() => {
      const connectionName = options.typeName || definitionClass.name;

      let edge: TypeExpression;
      const node = type(options.node!);
      const nodeTypeName = node.getTypeName({ storage, kind: 'output' });

      if (options.edge) {
        edge = type(options.edge);
      } else {
        const edgeTypeName = `${nodeTypeName}Edge`;
        defineEdge({ typeName: edgeTypeName, node })(null);
        edge = type(edgeTypeName);
      }

      storage.registerMetadata([
        new ObjectType({
          definitionClass,
          definitionName: connectionName,
          description: 'A connection to a list of items.'
        }),
        new Field({
          source: definitionClass,
          target: type(pageInfoType),
          fieldName: 'pageInfo',
          description: 'Information to aid in pagination.',
          args: [],
          resolver: bindStaticResolver(definitionClass, 'pageInfo') || defaultFieldResolver,
        }),
        new Field({
          source: definitionClass,
          target: List.of(edge),
          fieldName: 'edges',
          description: 'A list of edges.',
          args: [],
          resolver: bindStaticResolver(definitionClass, 'edges') || defaultFieldResolver,
        }),
      ]);
    });
  };
}

type DefineEdgeOptions = { typeName?: string, node: TypeArg | TypeExpression };

export function defineEdge(options: DefineEdgeOptions) {
  return function registerEdgeMetadata(definitionClass: Function | null) {
    const storage = getGlobalMetadataStorage();
    storage.deferRegister(() => {
      const node = type(options.node);
      const nodeTypeName = node.getTypeName({ storage, kind: 'output' });

      const edgeName = options.typeName
        || (definitionClass && definitionClass.name)
        || `${nodeTypeName}Edge`;

      storage.registerMetadata([
        new ObjectType({
          definitionClass,
          definitionName: edgeName,
          description: 'An edge in a connection.',
        }),
        new Field({
          source: definitionClass || edgeName,
          target: node,
          fieldName: 'node',
          description: 'The item at the end of the edge',
          args: [],
          resolver: definitionClass && bindStaticResolver(definitionClass, 'pageInfo') || defaultFieldResolver,
        }),
        new Field({
          source: definitionClass || edgeName,
          target: type('String'),
          fieldName: 'cursor',
          description: 'A cursor for use in pagination',
          args: [],
          resolver: definitionClass && bindStaticResolver(definitionClass, 'cursor') || defaultFieldResolver,
        }),
      ]);
    });
  };
}

const pageInfoType = new GraphQLObjectType({
  name: 'PageInfo',
  description: 'Information about pagination in a connection.',
  fields: () => ({
    hasNextPage: {
      type: new GraphQLNonNull(GraphQLBoolean),
      description: 'When paginating forwards, are there more items?'
    },
    hasPreviousPage: {
      type: new GraphQLNonNull(GraphQLBoolean),
      description: 'When paginating backwards, are there more items?'
    },
    startCursor: {
      type: GraphQLString,
      description: 'When paginating backwards, the cursor to continue.'
    },
    endCursor: {
      type: GraphQLString,
      description: 'When paginating forwards, the cursor to continue.'
    },
  })
});
