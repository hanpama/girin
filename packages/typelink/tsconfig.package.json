{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "./lib",
    "rootDir": "./src",
    "composite": true
  },
  "references": [
    {
      "path": "../connection/tsconfig.package.json"
    }
  ],
  "include": [
    "src"
  ],
  "exclude": [
    "tests"
  ]
}