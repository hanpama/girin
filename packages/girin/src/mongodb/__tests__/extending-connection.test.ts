import { typedef, gql, getGraphQLType } from "@girin/typelink";
import { printSchema, GraphQLSchema } from "graphql";

import { ModelConnection, field, Model } from "..";
import { Edge, ConnectionArguments } from "../../relay";


@typedef(gql`
  type Post {
    id: String!
    title: String!
  }
`)
class Post extends Model {
  @field('_id') id: string;
  @field() title: string;
}

@typedef(gql`
  type PostEdge {
    node: ${Post}
    cursor: String
  }
`)
class PostEdge extends Edge<Post> {}

@typedef(gql`
  type PostConnection {
    edges: ${PostEdge}
    pageInfo: PageInfo
    count: Int
  }
`)
class PostConnection extends ModelConnection<Post> {
  count() {
    return Post.getManager().collection.count(this.options.selector);
  }
}

@typedef(gql`
  type Query {
    posts(first: Int, last: Int, before: String, after: String): ${PostConnection}
  }
`)
class Query {
  static posts(source: null, args: ConnectionArguments) {
    return new PostConnection(args, {
      modelClass: Post,
      maxLimit: 20,
      sortOptions: [{
        fieldName: '_id',
        order: 1
      }]
    })
  }
}

describe('Extending connection', () => {

  it('generates schema as expected', () => {
    const schema = new GraphQLSchema({
      query: getGraphQLType(Query),
    });
    expect(printSchema(schema)).toMatchSnapshot();
  })
})