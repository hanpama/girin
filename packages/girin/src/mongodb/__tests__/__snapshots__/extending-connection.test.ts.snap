// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extending connection generates schema as expected 1`] = `
"type PageInfo {
  hasNextPage: Boolean
  hasPreviousPage: Boolean
  startCursor: String
  endCursor: String
}

type Post {
  id: String!
  title: String!
}

type PostConnection {
  edges: PostEdge
  pageInfo: PageInfo
  count: Int
}

type PostEdge {
  node: Post
  cursor: String
}

type Query {
  posts(first: Int, last: Int, before: String, after: String): PostConnection
}
"
`;
